File: rvc_service/RVCSDK/examples/Python/PointCloudCompensating.py
# Copyright (c) RVBUST, Inc - All rights reserved.
import PyRVC as RVC
import os
import numpy as np
import cv2


class CameraHelper:
    x2 = RVC.X2()
    x2_opts = None
    camera_id = RVC.CameraID_Left

    def __init__(self):
        RVC.SystemInit()

    def __del__(self):
        if self.x2.IsOpen():
            self.x2.Close()
        if self.x2.IsValid():
            RVC.X2.Destroy(self.x2)
        RVC.SystemShutdown()

    def Open(self):
        RVC.SystemInit()
        opt = RVC.SystemListDeviceTypeEnum.All
        ret, devices = RVC.SystemListDevices(opt)
        print("RVC Camera devices number:", len(devices))
        if len(devices) == 0:
            print("Can not find any RVC Camera!")
            RVC.SystemShutdown()
            return False
        if devices[0].IsFirmwareMatch() == False:
            print("device firmware mismatch, Please use RVCManager to upgrade the firmware")
            RVC.SystemShutdown()
            return False
        # Create and open RVC Camera
        device = devices[0]
        self.x2 = RVC.X2.Create(device)
        self.x2.Open()
        if self.x2.IsOpen() == False:
            print("Failed to open camera! Please check whether the camera is connected and make sure it is not occupied and supports X2.")
            RVC.X2.Destroy(self.x2)
            RVC.SystemShutdown()
            return False

        ret, info = device.GetDeviceInfo()
        self.camera_id = RVC.CameraID_Left
        if info.support_extra:
            self.camera_id = RVC.CameraID_Extra
        # Load capture options
        ret, self.x2_opts = self.x2.LoadCaptureOptionParameters()
        # Set the coordinate system to the camera coordinate system
        self.x2_opts.transform_to_camera = self.camera_id
        custom_transform_opt = RVC.X2_CustomTransformOptions()
        custom_transform_opt.coordinate_select = RVC.X2CustomTransformCoordinate.CoordinateSelect_Disabled
        ret = self.x2.SetCustomTransformation(custom_transform_opt)        
        return ret

    def Capture(self):
        is_capture_success = self.x2.Capture(self.x2_opts)
        if not is_capture_success:
            return RVC.PointMap(), RVC.Image()
        return self.x2.GetPointMap(), self.x2.GetImage(self.camera_id)

def CompensateForX2():
    camera_helper = CameraHelper()
    is_open_success = camera_helper.Open()
    if not is_open_success:
        print("RVC Camera fails to open!")
        return -1

    compensator = RVC.Compensator_FixedMarkers()

    # Capture a reference point map in static scene and initialize compensator.
    pm, image = camera_helper.Capture()
    ret = compensator.Initialize(pm, image, 0) # 0 for calibration board, 1 for concentric circles
    if ret != 0:
        print(f"Initialize fails! Error code: {ret}.")
        return -1
    print("Initialize successfully!")
    
    save_address = "Data"
    os.makedirs(save_address, exist_ok=True)
    # It is recommended to update the model within every 50 captures.
    # The value 50 is not fixed and needs to take into account the speed of the ambient temperature change and the camera shooting rhythm.
    # In this sample program, there are only 10 frames in between as a demonstration.
    interval_between_compensation = 10  # 100
    pm, image = camera_helper.Capture()
    pm.SaveWithImage(f"{save_address}/pointcloud_0.ply", image, RVC.PointMapUnitEnum.Meter, True)
    image.SaveImage(save_address + "/image_0.png")
    for _ in range(interval_between_compensation):
        camera_helper.Capture()
    
    # Capture the fixed markers and update compensation model.
    pm, image = camera_helper.Capture()
    ret, drift_distance = compensator.Update(pm, image)
    if ret != 0:
        print(f"Update compensation model fails! Error code: {ret}.")
        return -1
    print(f"Current drift distance: {drift_distance}")
    print("Update compensation model successfully!")
    
    # After update compensation model, we can compensate for the point cloud.
    pm, image = camera_helper.Capture()
    pm.SaveWithImage(f"{save_address}/pointcloud_n_before_compensate.ply", image, RVC.PointMapUnitEnum.Meter, True)
    image.SaveImage(save_address + "/image_n.png")
    ret, pm = compensator.Apply(pm)
    pm.SaveWithImage(f"{save_address}/pointcloud_n_after_compensate.ply", image, RVC.PointMapUnitEnum.Meter, True)
    print("Compensate pm successfully!")

    return 0


if __name__ == "__main__":
    """
    Use case: Compensate for the point cloud drift caused by changes of camera or ambient temperature
    
    Process: 
    1. Capture the fixed markers to initialize; 
    2. Use the camera normally and take n frames;
    3. Keep the camera in the same position as it was initialized and capture the same fixed markers to update the model; 
    4. Continue to use the camera to take n frames, and for each frame we can apply the updated model to compensate;
    5. Repeat the process of 3 and 4.
    It is recommended that n should be within 50. And the initialization step is recommended to be done after hand-eye calibration.
    """
    CompensateForX2()

File: rvc_service/RVCSDK/examples/CSharp/PointCloudCompensating.cs
/**
 * Important:
 * When the capture mode is set to `CaptureMode_SwingLineScan`,
 * `correspond2d` MUST be enabled (set to true), and
 * the option "capture 2D image during 3D acquisition" must also be enabled.
 * Otherwise, the sample will fail.
 *
 * 注意：
 * 使用摆动线扫 (SwingLineScan) 执行示例时，必须同时开启：
 *   1. 2D 对齐（correspond2d = true）
 *   2. 在 3D 采集中拍摄 2D 图选项
 * 否则示例会执行失败。
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RVC_CSharp;


namespace RVC_CSharp
{

    class Program
    {
        static void Main(string[] args)
        {
            //CompensateForX2Extra();
            CompensateForX2();
            //CompensateForX1();
            Console.ReadKey();

        }

        static void MakeDirectories(string path)
        {
            if (!Directory.Exists(path))
                Directory.CreateDirectory(path);
        }

        #region X2Extra Implementation
        class CameraHelperX2Extra : IDisposable
        {
            private X2 m_x2;
            private X2.CaptureOptions m_opts;

            public CameraHelperX2Extra()
            {
                System.Init();
            }

            public bool Open()
            {
                // List Devices
                var devices = System.ListDevices();
                if (devices.Count == 0) return false;
                if(devices[0].IsFirmwareMatch() == false)
                {
                    Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return false;
                }
                // Create and open camera
                m_x2 = X2.Create(devices[0]);
                if (!m_x2.Open()) return false;

                // Load capture options
                if (!m_x2.LoadCaptureOptionParameters(ref m_opts)) return false;

                // Configure transform
                m_opts.transform_to_camera = CameraID.CameraID_Extra;
                var customTransformOpt = X2.CustomTransformOptions.Default();
                return m_x2.SetCustomTransformation(customTransformOpt);
            }

            public void Capture(out PointMap pointMap, out Image image)
            {
                if (!m_x2.Capture(m_opts))
                {
                    Console.WriteLine($"Failed to Capture .Error: {System.GetLastErrorMessage()} ");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                }

                pointMap = m_x2.GetPointMap();
                image = m_x2.GetImage(CameraID.CameraID_Extra);

            }

            public void Dispose()
            {
                if (m_x2.IsOpen()) m_x2.Close();
                if (m_x2.IsValid()) X2.Destroy(m_x2);
                System.Shutdown();
            }
        }

        static void CompensateForX2Extra()
        {
            using var camera = new CameraHelperX2Extra();
            if (!camera.Open())
            {
                Console.WriteLine("RVC X2 Camera fails to open!");
                Console.ReadKey();
                return;
            }

            using var compensator = new Compensator();

            // 1. Initialize
            camera.Capture(out PointMap pm, out Image image);

            try
            {
                compensator.Initialize(pm, image, CompensatorMarkerType.MarkerType_CalibrationBoard);
                Console.WriteLine("Initialize successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Initialize failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 2. Update
            var dir = Path.Combine(Environment.CurrentDirectory, "Data");
            MakeDirectories(dir);

            int interval = 10;
            for (int i = 0; i < interval; i++)
            {
                camera.Capture(out PointMap tmpPm, out Image tmpImg); // Skip frames
            }

            camera.Capture(out PointMap updatePm, out Image updateImg);

            try
            {
                var drift = compensator.Update(updatePm, updateImg);
                Console.WriteLine($"Drift distance: {drift:F4}m\nUpdate success!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Update failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 3. Apply
            camera.Capture(out PointMap finalPm, out Image finalImg);
            finalPm.Save(Path.Combine(dir, "before_compensate.ply"), PointMapUnit.Millimeter, true, finalImg);
            compensator.Apply(finalPm);
            finalPm.Save(Path.Combine(dir, "after_compensate.ply"), PointMapUnit.Millimeter, true, finalImg);
        }
        #endregion

        #region X2 Implementation
        class CameraHelperX2 : IDisposable
        {
            private X2 m_x2;
            private X2.CaptureOptions m_opts;

            public CameraHelperX2()
            {
                System.Init();
            }

            public bool Open()
            {
                // List Devices
                var devices = System.ListDevices();
                if (devices.Count == 0) return false;
                if(devices[0].IsFirmwareMatch() == false)
                {
                    Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return false;
                }
                // Create and open camera
                m_x2 = X2.Create(devices[0]);
                if (!m_x2.Open()) return false;

                // Load capture options
                if (!m_x2.LoadCaptureOptionParameters(ref m_opts)) return false;

                // Configure transform
                m_opts.transform_to_camera = CameraID.CameraID_Left;
                var customTransformOpt = X2.CustomTransformOptions.Default();
                return m_x2.SetCustomTransformation(customTransformOpt);
            }

            public void Capture(out PointMap pointMap, out Image image)
            {
                if (!m_x2.Capture(m_opts))
                {
                    Console.WriteLine($"Failed to Capture .Error: {System.GetLastErrorMessage()} ");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                }

                pointMap = m_x2.GetPointMap();
                image = m_x2.GetImage(CameraID.CameraID_Left);

            }

            public void Dispose()
            {
                if (m_x2.IsOpen()) m_x2.Close();
                if (m_x2.IsValid()) X2.Destroy(m_x2);
                System.Shutdown();
            }
        }

        static void CompensateForX2()
        {
            using var camera = new CameraHelperX2();
            if (!camera.Open())
            {
                Console.WriteLine("RVC X2 Camera fails to open!");
                Console.ReadKey();
                return;
            }

            using var compensator = new Compensator();

            // 1. Initialize
            camera.Capture(out PointMap pm, out Image image);

            try
            {
                compensator.Initialize(pm, image, CompensatorMarkerType.MarkerType_CalibrationBoard);
                Console.WriteLine("Initialize successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Initialize failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 2. Update
            var dir = Path.Combine(Environment.CurrentDirectory, "Data");
            MakeDirectories(dir);

            int interval = 10;
            for (int i = 0; i < interval; i++)
            {
                camera.Capture(out PointMap tmpPm, out Image tmpImg); // Skip frames
            }

            camera.Capture(out PointMap updatePm, out Image updateImg);

            try
            {
                var drift = compensator.Update(updatePm, updateImg);
                Console.WriteLine($"Drift distance: {drift:F4}m\nUpdate success!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Update failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 3. Apply
            camera.Capture(out PointMap finalPm, out Image finalImg);
            finalPm.Save(Path.Combine(dir, "before_compensate.ply"), PointMapUnit.Millimeter, true, finalImg);
            compensator.Apply(finalPm);
            finalPm.Save(Path.Combine(dir, "after_compensate.ply"), PointMapUnit.Millimeter, true, finalImg);
        }
        #endregion

        #region X1 Implementation
        class CameraHelperX1 : IDisposable
        {
            private X1 m_x1;
            private X1.CaptureOptions m_opts;

            public CameraHelperX1() => System.Init();

            public bool Open()
            {
                var devices = System.ListDevices();
                if (devices.Count == 0) return false;
                if(devices[0].IsFirmwareMatch() == false)
                {
                    Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                    return false;
                }
                m_x1 = X1.Create(devices[0], CameraID.CameraID_Left);
                if (!m_x1.Open()) return false;

                if (!m_x1.LoadCaptureOptionParameters(ref m_opts)) return false;

                m_opts.transform_to_camera = true;
                //X1.CustomTransformOptions custom_transform_opt;
                //custom_transform_opt.coordinate_select = X1.CustomTransformOptions.CoordinateSelect.CoordinateSelect_Camera;
                return true;
            }


            public void Capture(out PointMap pointMap, out Image image)
            {
                if (!m_x1.Capture(m_opts))
                {
                    Console.WriteLine($"Failed to Capture .Error: {System.GetLastErrorMessage()} ");
                    RVC_CSharp.System.Shutdown();
                    Console.ReadKey();
                }

                pointMap = m_x1.GetPointMap();
                image = m_x1.GetImage();

            }

            public void Dispose()
            {
                if (m_x1.IsOpen()) m_x1.Close();
                if (m_x1.IsValid()) X1.Destroy(m_x1);
                System.Shutdown();
            }
        }

        static void CompensateForX1()
        {
            using var camera = new CameraHelperX1();
            if (!camera.Open())
            {
                Console.WriteLine("RVC X1 Camera fails to open!");
                Console.ReadKey();
                return;
            }

            using var compensator = new Compensator();

            // 1. Initialize
            camera.Capture(out PointMap pm, out Image image);

            try
            {
                compensator.Initialize(pm, image, CompensatorMarkerType.MarkerType_CalibrationBoard);
                Console.WriteLine("Initialize successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Initialize failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 2. Update
            var dir = Path.Combine(Environment.CurrentDirectory, "Data");
            MakeDirectories(dir);

            int interval = 10;
            for (int i = 0; i < interval; i++)
            {
                camera.Capture(out PointMap tmpPm, out Image tmpImg); // Skip frames
            }

            camera.Capture(out PointMap updatePm, out Image updateImg);

            try
            {
                var drift = compensator.Update(updatePm, updateImg);
                Console.WriteLine($"Drift distance: {drift}m\nUpdate success!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Update failed: {ex.Message}");
                Console.ReadKey();
                return;
            }

            // 3. Apply
            camera.Capture(out PointMap finalPm, out Image finalImg);
            finalPm.Save(Path.Combine(dir, "x1_before.ply"), PointMapUnit.Millimeter, true, finalImg);
            compensator.Apply(finalPm);
            finalPm.Save(Path.Combine(dir, "x1_after.ply"), PointMapUnit.Millimeter, true, finalImg);
        }
        #endregion
    }

}